geopandas/geodataframe.py

```diff
@@ -392,6 +524,8 @@ def merge(self, *args, **kwargs):
         if isinstance(result, DataFrame) and geo_col in result:
             result.__class__ = GeoDataFrame
             result.crs = self.crs
+            g = vectorized.VectorizedGeometry(result[geo_col])
+            result[geo_col] = list(g)
```

Comment:
The list here causes a densification. I think we can avoid that by passing the `result` to GeoDataFrame and passing `g` as geometry to that (then the block reconstruction of the `__init__` is reused). We might also make a helper function for things like this that is used both here and in the `__init__`